### PR TITLE
fix(revive): fail System.terminate when deferred destruction cannot burn ED

### LINKED ISSUE #13017: [pallet-revive] System termination can succeed when deferred destruction fails
### Is there an existing issue?

- [x] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [x] This is not a support question.

## Summary

`System.terminate()` can successfully schedule contract destruction and perform its immediate beneficiary transfer, while the deferred destruction later fails and the failure is silently discarded.

The destruction schedule is consumed before finalization, and the result of `do_terminate()` is converted to `.ok()`. If `do_terminate()` then fails, for example because a native balance lock/freeze makes the contract ED non-reducible, no retry state remains and the contract stays live and callable even though the System precompile returned success and the enclosing contract execution succeeded.

Relevant flow:

- [`System::terminate` returns success after calling `terminate_caller`](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/precompiles/builtin/system.rs#L101-L107)
- [`terminate_caller` immediately transfers the reducible balance and schedules destruction](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/exec.rs#L2560-L2589)
- [first-frame finalization consumes `contracts_to_be_destroyed` with `mem::take` and discards `do_terminate()` errors with `.ok()`](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/exec.rs#L1699-L1712)
- [`do_terminate()` is fallible and rolls back only its own finalizer-local storage transaction](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/exec.rs#L1799-L1871)
- [native `destroy_contract()` burns the contract ED exactly with `Fortitude::Polite`](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/deposit_payment.rs#L168-L180)

As a result, a successful call can leave this observable state:

- the beneficiary already received the contract's reducible balance;
- the termination request has been consumed;
- the contract still has its authoritative contract state and code;
- the contract remains callable;
- there is no persistent pending-termination state or retry.

## Why this seems inconsistent with the intended semantics

The public System interface documents that `terminate` may revert when balance locks prevent termination:

[`ISystem.sol`: "the contract introduced balance locks"](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/uapi/sol/ISystem.sol#L59-L65)

However, when the relevant failure occurs during deferred `do_terminate()`, it is currently hidden instead of making the enclosing termination operation fail.

There is also relevant historical context:

- [#9699](https://github.com/paritytech/polkadot-sdk/pull/9699) introduced deferred destruction and contained older discussion where final termination failure could be treated as a non-bubbling no-op. At that point, however, the beneficiary transfer was still part of final cleanup rather than an already-committed immediate effect.
- [#10302](https://github.com/paritytech/polkadot-sdk/pull/10302) changed termination so the free/reducible balance is transferred immediately while ED/storage cleanup remains deferred, and introduced the current `do_terminate(...).ok()` finalization shape.
- [#11847](https://github.com/paritytech/polkadot-sdk/pull/11847) introduced the current `Deposit::destroy_contract()` path, including the exact `Fortitude::Polite` ED burn that can legitimately fail when external balance constraints exist.
- [#12144](https://github.com/paritytech/polkadot-sdk/pull/12144) fixed another instance of the same finalizer-error swallowing pattern: a Root-origin path could make `do_terminate()` fail, the error was discarded by the existing finalizer, and `ContractInfo` remained on-chain. The PR fixed that specific failure trigger while the generic `.ok()` handling remained.

The regression added with that fix still documents the previous silent failure:

[`ContractInfo` stayed put because `do_terminate` silently failed](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/tests/sol/contract.rs#L862-L864)

## Expected behavior / design question

The balance lock itself preventing destruction seems valid. I don't think termination should bypass external locks/freezes.

The question is how a late finalizer failure should affect the already accepted System termination.

The behavior that currently seems most consistent with the post [#10302](https://github.com/paritytech/polkadot-sdk/pull/10302) API is that a late `System.terminate()` finalizer failure should become visible to the enclosing operation rather than leaving a normally live contract after a successful termination path.

Before implementing a fix for this, I would like to confirm the intended semantics:

1. Should a deferred `System.terminate()` finalizer failure revert the enclosing top-level contract execution?
2. If several contracts are scheduled for destruction and one finalizer fails, should the scheduled destruction set be atomic?
3. Should EVM/PVM same-transaction destruction use the same failure policy, or should System termination have distinct failure semantics?

One possible implementation direction would be to extend the relevant first-frame storage transaction through scheduled-destruction finalization. That would preserve the immediate beneficiary-transfer semantics on success, while rolling back the transfer and other execution state if final destruction fails, while preserving consumed gas/weight accounting.

I have not implemented that change yet because it affects termination transaction semantics and I would prefer to confirm the intended behavior first.

### Steps to reproduce

A minimal reproduction is:

1. Create and fund a revive contract.
2. Make part of its native balance non-reducible using a balance lock/freeze, such that the contract ED cannot later be burned with `Fortitude::Polite`.
3. From the contract, call the System precompile `terminate(beneficiary)`.
4. Complete the enclosing contract call successfully.

During finalization:

- the immediate `Preservation::Preserve` transfer has already moved the currently reducible balance to the beneficiary;
- `do_terminate()` reaches native `destroy_contract()`;
- the exact polite ED burn fails because the remaining balance is non-reducible;
- `do_terminate()` rolls back its local storage transaction and returns the error;
- the deferred-finalization caller discards that error with `.ok()` after `contracts_to_be_destroyed` has already been consumed.

Observed final state:

- the beneficiary keeps the immediate transfer;
- the top-level call reports success;
- the contract still exists;
- its code and authoritative contract state remain;
- the contract remains callable;
- no destruction request remains to retry later.

The relevant error-discarding call site is:

[`do_terminate(...).map(|_| ()).ok()`](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/exec.rs#L1705-L1712)

and the finalizer-local rollback is here:

[`do_terminate` storage transaction](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/revive/src/exec.rs#L1857-L1871)

### Reachability of the balance constraint

This strep 2 is not limited to a manually constructed test state. A contract account can
legitimately acquire native balance constraints from other runtime components.

For example, Asset Hub Westend currently exposes the Vesting precompile to
`pallet-revive` contracts:

[`VestingPrecompile<Self>` in the revive runtime configuration](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L1387-L1435)

A revive contract can call `vestedTransfer(address(this), ...)`; the precompile
dispatches `pallet_vesting::vested_transfer` with the contract account as the
signed caller:

[`VestingPrecompile::vested_transfer`](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/vesting/precompiles/src/lib.rs#L168-L215)

The resulting vesting schedule installs a native balance lock on the target
account:

[`write_lock` updates the Balances lock](https://github.com/paritytech/polkadot-sdk/blob/a08ee7b2b89e3cf491ebf6a3e366751a40d883eb/substrate/frame/vesting/src/lib.rs#L620-L628)

This gives a concrete runtime path where the contract has non-reducible native
balance when deferred termination reaches `destroy_contract()`.

The issue is not specific to Vesting: any legitimate runtime mechanism that
makes a fallible `do_terminate()` operation fail after termination was scheduled
can expose the same finalizer-error handling problem.

### PR DESCRIPTION
**Pull request by dimartiro**
# Description

Fixes #13017.

`System.terminate` returned success after `terminate_caller` moved the reducible balance and scheduled destruction, even when the deferred `do_terminate` later failed. The first-frame finalizer consumed `contracts_to_be_destroyed` with `mem::take` and discarded the error with `.ok()`. If the polite ED burn then failed — for example because a native freeze/lock made the remaining balance non-reducible — the beneficiary kept the transfer, the contract stayed live and callable, and nothing was left to retry.

`ISystem.sol` documents that `terminate` reverts when the contract introduced balance locks. That failure was hidden when it happened in the deferred finalizer rather than in the precompile itself.

## The change

Scheduled destruction now runs **inside the first-frame storage transaction**, so a late `do_terminate` failure aborts the enclosing top-level call and rolls back the immediate beneficiary transfer. Same policy for `System.terminate` and same-tx `SELFDESTRUCT`: they share this finalizer.

- `Stack::finalize_scheduled_destructions` is called from `run` at the end of the first frame, before that frame's `with_transaction` commits.
- `do_terminate` returns the storage-deposit refund instead of applying `transaction_meter.terminate` itself. Meter updates run only after every scheduled destroy succeeds, so a mid-batch failure cannot leave the deposit meter thinking a still-live contract was reaped.
- After a successful destroy of the first-frame account, its cached `ContractInfo` is invalidated so `pop_frame` does not write it back.

Locks/freezes are not bypassed. Gas/weight consumed by the reverted call is still charged via `pop_frame(persist=false)`.

## Integration

No API change. Downstream runtimes pick this up on the next `pallet-revive` bump. Observable behaviour: a `System.terminate` / same-tx `SELFDESTRUCT` whose deferred destruction cannot burn the ED now fails the enclosing call and restores balances, instead of reporting success with a live contract.

## Review Notes

The historical `.ok()` dates to [#10302](https://github.com/paritytech/polkadot-sdk/pull/10302), when the beneficiary transfer moved from final cleanup to an immediate `Preservation::Preserve` send. [#11847](https://github.com/paritytech/polkadot-sdk/pull/11847) made the ED burn `Fortitude::Polite`, which can legitimately fail. [#12144](https://github.com/paritytech/polkadot-sdk/pull/12144) fixed one swallowed-error trigger (Root origin) without changing the generic `.ok()`.

This implements the direction in the issue: extend the first-frame storage transaction through scheduled-destruction finalization. Several scheduled destroys in one tx are atomic with that transaction.

Deliberately left: bypassing locks/freezes so terminate always reaps. The public interface says those constraints should revert.

## Testing

- `precompile_terminate_reverts_when_ed_is_frozen`: native freeze of `reserved + ED`, then `System.terminate`. Enclosing call is `Err`, contract stays live, extra balance is restored, beneficiary does not keep the transfer.
- Existing sol terminate tests (20 solc), including first-frame `System.terminate` actually destroying the contract, rollback if a later frame reverts, and same-tx vs prior-tx `SELFDESTRUCT`.
- PVM `can_self_destruct_while_live` / `self_destruct_by_precompile_works` and `destroy_contract_reaps_account_and_clears_native_deposit_map`.

# Checklist

* [x] My PR includes a detailed description as outlined in the "Description" and its two subsections above.
* [x] My PR follows the [labeling requirements](https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/CONTRIBUTING.md#Process) of this project (at minimum one label for `T` required)
* [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: substrate/frame/revive/src/exec.rs [hunk 1]
```diff
@@ -1501,11 +1501,22 @@ where
...
+					.finalize(contract)
+					.map_err(|e| ExecError { error: e, origin: ErrorOrigin::Callee })?;
+			}
+
+			// Deferred destruction must run inside this storage transaction so a
+			// late `do_terminate` failure rolls back the enclosing execution,
+			// including the immediate beneficiary transfer. Reverts return above,
+			// so we never destroy after a reverted callee.
+			if is_first_frame {
+				self.finalize_scheduled_destructions()
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs]: This is the part I don't want to take: a contract that is already off the stack failing the first frame. Reasoning in the review body.

File: substrate/frame/revive/src/tests/sol/terminate.rs [hunk 2]
```diff
@@ -698,3 +707,77 @@ fn call_after_terminate_works(fixture_type: FixtureType, method: u8) {
...
+				Terminate::terminateCall {
+					method: METHOD_PRECOMPILE,
+					beneficiary: DJANGO_ADDR.0.into(),
+				}
+				.abi_encode(),
+			)
+			.build();
+
+		assert!(
+			result.result.is_err(),
```
Comment (athei) [inline substrate/frame/revive/src/tests/sol/terminate.rs]: With the check moved into `terminate_caller` this becomes a revert of the precompile call, not of the top-level call. Assert on that plus `get_balance(&DJANGO) == django_before`. The contract staying live and keeping its balance then follows from the revert.

**Review by athei — CHANGES_REQUESTED** [review]
Thanks for the analysis, it matches what I found when looking at this. I don't want to take the transaction boundary change though, for two reasons.

The destruction runs after the whole stack has unwound, so this lets a callee fail its caller after the caller already saw a successful return. A Solidity `try/catch` around the call cannot observe that. The comment you removed in `do_terminate` (the contract that called SELFDESTRUCT is no longer on the call stack, so we cannot fail here) was the design decision behind the `.ok()`, not an oversight. It also makes SELFDESTRUCT fallible, which it never is in the EVM.

It also collides with #13039, which is queued for merge. That PR settles the other failure behind the same `.ok()`, an unaffordable beneficiary ED, as skip and hold: the termination is skipped and everything else in the call stays committed and backed. With this change that case reverts the whole transaction instead, and `tests/sol/terminate_storage.rs` from #13039 fails.

So the deferred loop stays non-failing. What is actually broken is the promise in `ISystem.sol` that `terminate` reverts when the contract has balance locks: `terminate_caller` never checks that. Please move the fix there. After the immediate transfer, check that the polite ED burn in `destroy_contract` can still succeed once `refund_all` has released our holds, and fail the precompile call with a dedicated error if not. In pallet_balances terms the burn needs `free + reserved - frozen >= ED`, and releasing holds does not change `frozen`, so the check can be done at terminate time. If the `fungible::Inspect` bounds on `T::Currency` are too coarse to compute that exactly, a stricter check that rejects any foreign freeze on the account is fine, that is what `ISystem.sol` documents anyway. This fails synchronously inside the precompile, where the calling contract can catch it, and a third party locking a contract via vested transfer becomes a visible error instead of a silent no-op. A lock added by a later frame in the same transaction still ends in the silent skip; that is the same residual #13039 accepts for the deposit limit and is fine.

Keep the freeze test, but assert that the precompile call reverts and the beneficiary balance is unchanged, instead of asserting that the top-level call fails. Leave the SELFDESTRUCT opcode alone. Rebase onto master once #13039 is in, and add the prdoc and a T7 label.

**Comment by dimartiro** [comment]
@athei 

You're right that the removed comment was the design decision and not an oversight. Dropped the transaction-boundary change: the deferred loop and the `SELFDESTRUCT` path are back to their original form, and the check now sits in `terminate_caller` after the immediate transfer.

Your two options turn out to be one, and neither needs new bounds. `reducible_balance(Expendable, Polite)` *is* `free + reserved - frozen` clamped at zero, so the existing `fungible::Inspect` bound computes your formula exactly. And since the transfer leaves `free = max(max(frozen - reserved, 0), ED)`, that formula rejects precisely when `frozen > reserved` — i.e. it coincides with the strict "reject any foreign freeze" fallback:

```rust
ensure!(
    T::Currency::reducible_balance(&parent_account_id, Preservation::Expendable, Fortitude::Polite)
        >= T::Currency::minimum_balance(),
    Error::<T>::TerminateBalanceLocked,
);
```

I also had to add the new error to `Error::try_to_revert`, otherwise it traps instead of being catchable, which was the point of moving the check here.

One residual: a freeze with `frozen <= reserved` is invisible to this, but the burn really needs `frozen <= reserved - our_holds`, so that band still ends in the silent skip. Closing it exactly means summing `Deposit::total_on_hold` at terminate time — happy to, though it looks like the same class of residual #13039 accepts.

Test now asserts the revert, its message, and that the beneficiary balance is unchanged. prdoc added, bumped `minor` since the new `Error` variant is public API — say if you'd rather keep it `patch`. Will rebase once #13039 lands; it doesn't touch `exec.rs`, so I don't expect a conflict.

File: substrate/frame/revive/src/exec.rs [hunk 3]
```diff
@@ -2712,6 +2712,24 @@ where
...
+		// The deferred destruction cannot fail: the contract is no longer on the call stack by
+		// then, so its error is swallowed. Reject here instead the one failure we can still see:
+		// a freeze that pins the ED and would make the `Polite` burn in `destroy_contract` fail.
+		//
+		// `refund_all` only lowers `reserved` and never touches `frozen`, so a burn that does not
+		// fit now will not fit later either. Checking after the transfer above is what makes this
+		// exact: `Preservation::Preserve` leaves exactly `max(frozen - reserved, ED)` behind, so
+		// the reducible balance is below the ED precisely when a foreign freeze exceeds our holds.
+		let ed = T::Currency::minimum_balance();
+		ensure!(
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs:2724]: This rejects every terminate, not only frozen ones. Every live contract carries the consumer ref added at instantiation (only removed in `do_terminate`) and has a single provider, so `can_dec_provider` is false and `reducible_balance(Expendable, Polite)` adds the ED to the untouchable part exactly like `Preserve` does. After the `Preserve` transfer above, `free = max(frozen - reserved, ED)`, and this computes `free - max(frozen - reserved, ED) = 0 < ED` for every contract. CI shows it: `base_case`, `terminate_twice`, `call_after_terminate_works`, `pvm::self_destruct_by_precompile_works`, the lazy removal tests and others fail with this error's revert string.

File: substrate/frame/revive/src/exec.rs [hunk 4]
```diff
@@ -2712,6 +2712,24 @@ where
 			&self.exec_config,
 		)?;
 
+		// The deferred destruction cannot fail: the contract is no longer on the call stack by
+		// then, so its error is swallowed. Reject here instead the one failure we can still see:
+		// a freeze that pins the ED and would make the `Polite` burn in `destroy_contract` fail.
+		//
+		// `refund_all` only lowers `reserved` and never touches `frozen`, so a burn that does not
```
Comment (athei) [inline substrate/frame/revive/src/exec.rs:2719]: The direction is reversed. The native `refund_on_hold` is a `transfer_on_hold` with `Restriction::Free` to the depositor, so the contract's `free` stays put while `reserved` drops, and the polite untouchable `frozen - reserved` grows. In `do_terminate`, after `refund_all` and `dec_consumers`, any nonzero freeze or leftover foreign hold keeps the balances consumer, so the burn sees `untouchable = max(frozen - holds_left, ED)` against `free = ED` and fails. A freeze of 1 passes the formula here and still ends in the silent skip. Please rewrite this comment for whatever the check becomes, and drop "exact" unless it is.

File: substrate/frame/revive/src/lib.rs [hunk 5]
```diff
@@ -666,6 +666,13 @@ pub mod pallet {
 		/// `seal_terminate` was invoked on an EIP-7702 delegated EOA. Delegated accounts
 		/// cannot be torn down via the contract-termination path.
 		CannotTerminateDelegatedAccount = 0x44,
+		/// A balance freeze pins the contract's existential deposit, so terminating it would
+		/// not be able to burn that ED.
+		///
+		/// The actual destruction is deferred to the end of the call stack, where it can no
```
Comment (athei) [inline substrate/frame/revive/src/lib.rs:672]: The deferred destruction can still fail, its error is just not visible to anyone. Please say that, and make the first line describe the condition the check ends up enforcing (a freeze or foreign hold on the account, not a freeze that "pins the ED"). Also the check runs after the immediate transfer, so "up-front" is misleading.

File: substrate/frame/revive/src/tests/sol/terminate.rs [hunk 6]
```diff
@@ -698,3 +707,69 @@ fn call_after_terminate_works(fixture_type: FixtureType, method: u8) {
...
+				.abi_encode(),
+			)
+			.build_and_unwrap_contract();
+
+		let _ = <Test as Config>::Currency::set_balance(&account_id, min_balance + extra);
+		// Pin ED *and* current reserved: polite `reducible_balance` subtracts
+		// reserved from frozen, so freezing only the ED is a no-op while holds
+		// remain. `PGasDeposit::destroy_contract` thaws the PGAS freeze itself,
+		// so the pin has to be a native Balances freeze.
+		let pin = Balances::reserved_balance(&account_id).saturating_add(min_balance);
```
Comment (athei) [inline substrate/frame/revive/src/tests/sol/terminate.rs:748]: This test cannot tell a correct check from one that rejects everything, which is what the current one does. Please add:
- a control in the same setup: without the freeze the same call destroys the contract
- the boundary that matters: a freeze at or below the reserved amount (e.g. 1), which must revert too, and a foreign hold if the check covers holds
- assertions that the contract is still live and its balance unchanged after the revert, as the description claims
- one case where a Solidity caller catches the revert and continues, since being catchable is why the check lives in the precompile

Use a foreign lock (`Balances::set_lock` with a non-revive `LockIdentifier`) or a freeze reason pallet-revive does not own instead of `PGasMinBalance`; that also covers the lock half of the ISystem.sol wording. The comment above says freezing only the ED is a no-op while holds remain. That is true for this check but false for the deferred burn, which is the case the test exists for.

File: substrate/frame/revive/uapi/sol/ISystem.sol [hunk 7]
```diff
@@ -61,7 +61,7 @@ interface ISystem {
 	/// - called from constructor
 	/// - called from static context
 	/// - called from delegate context
-	/// - the contract introduced balance locks
+	/// - a balance lock or freeze on the contract pins its existential deposit
```
Comment (athei) [inline substrate/frame/revive/uapi/sol/ISystem.sol:64]: Once the check is fixed this should state the actual condition, e.g. "the contract's balance has a lock, freeze or hold that terminate cannot release".

File: prdoc/pr_13085.prdoc [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    moved the reducible balance to the beneficiary and scheduled the destruction, but the deferred
+    destruction runs at the end of the call stack, where the contract is no longer on the stack and
+    its error is discarded. If the polite existential-deposit burn then failed — for example because
+    a native freeze made the remaining balance non-reducible — the beneficiary kept the transfer,
+    the contract stayed live and callable, and nothing was left to retry.
+
+    `terminate_caller` now rejects that case up-front: after the immediate transfer it checks that
+    the contract's reducible balance still covers the existential deposit, and otherwise fails with
+    the new `Error::TerminateBalanceLocked`. Refunding the storage deposit only lowers `reserved` and
+    never touches `frozen`, so a burn that does not fit at terminate time will not fit later either.
```
Comment (athei) [inline prdoc/pr_13085.prdoc:16]: This is the reasoning that does not hold (see the comment in `terminate_caller`). Please rewrite once the check is fixed: say under which condition `terminate` now reverts, and that contracts that used to get a silent no-op in that case now see a revert they can catch. Also no em dashes (lines 9 and 10).

**Review by athei — CHANGES_REQUESTED** [review]
The check is in the right place now, but as written it rejects every `System.terminate`, and test-linux-stable is red on about 40 pallet-revive terminate tests because of it. Details inline.

On the residual you asked about: the `frozen <= reserved` band is not an edge case, it is where the bug lives. After `refund_all` our hold is gone, so any nonzero freeze, and any hold under another reason, makes the polite ED burn fail. The condition the deferred destruction actually needs is no freeze at all and no hold other than `StorageDepositReserve`.

`fungible::Inspect` cannot read `frozen`, and I'd rather not add a `Currency` bound just for this. The simplest exact option I see is a dry run: after the transfer in `terminate_caller`, run the balance part of the teardown (`refund_all`, `dec_consumers`, `destroy_contract`) inside a storage transaction that is always rolled back, and fail with `TerminateBalanceLocked` if it errors. Factor that part out of the `delete_contract` closure in `do_terminate` so both paths share it instead of copying it. The final transfer can stay out of it since it moves nothing at terminate time. This covers both deposit backends, holds, freezes and the PGAS side without balance arithmetic in exec.rs. The precompile currently charges `seal_terminate_logic` once; the dry run repeats most of that work, so it needs to be charged as well. If you see a cheaper way that is exact, fine by me.

Please rebase onto master now that #13039 is in. `tests/sol/terminate_storage.rs` calls `System.terminate` and has to stay green with the fixed check.

The PR title and description still describe the first approach (first-frame transaction, `finalize_scheduled_destructions`, `do_terminate` returning the refund, "No API change"). Please update both to the current diff.

`major` for pallet-revive is right since `Error` is an exhaustive public enum. `patch` for uapi is fine.

File: substrate/frame/revive/src/exec.rs (hunk 3 above)
Comment (dimartiro) [reply substrate/frame/revive/src/exec.rs]: Right, the formula could never pass once `Preserve` had left `free` at the untouchable floor. Replaced it with the dry run you suggested: `terminate_caller` runs `release_contract_balance` (`refund_all`, `dec_consumers`, `destroy_contract`, factored out of `do_terminate` so both paths share it) inside a `with_transaction` that always rolls back, and fails with `TerminateBalanceLocked` if it errors. `base_case`, `terminate_twice`, `call_after_terminate_works` and the rest pass again; full `pallet-revive` lib suite is green locally. The dry run is charged through a new `RuntimeCosts::TerminateDryRun` (`seal_terminate_logic` again). 1a602ef9d89

File: substrate/frame/revive/src/exec.rs (hunk 4 above)
Comment (dimartiro) [reply substrate/frame/revive/src/exec.rs]: Rewrote the comment for the dry run: it states the condition (a freeze or lock, or a hold other than the storage deposit), no longer claims exactness, and notes the residual of a lock added by a later frame. 1a602ef9d89

File: substrate/frame/revive/src/lib.rs (hunk 5 above)
Comment (dimartiro) [reply substrate/frame/revive/src/lib.rs]: Done. First line is now the condition ("a freeze, lock or hold that terminating it cannot release"), it says the deferred destruction can still fail but nobody sees the error, and "up-front" is gone: it now says the dry run happens after the immediate transfer. 1a602ef9d89

File: substrate/frame/revive/src/tests/sol/terminate.rs (hunk 6 above)
Comment (dimartiro) [reply substrate/frame/revive/src/tests/sol/terminate.rs]: Replaced it with `precompile_terminate_with_encumbered_balance` (Solc and Resolc) over: no encumbrance as the control (contract destroyed, beneficiary paid), a foreign lock of 1 (the contract has 872 native reserved, so this is inside the `frozen <= reserved` band), a foreign lock of 1,000,000, and a hold of 1 under `HoldReason::AddressMapping`. Each encumbered case asserts the revert message, that the contract is still live, and that contract and beneficiary balances are unchanged. Locks use `Balances::set_lock` with a non-revive id. `precompile_terminate_revert_can_be_caught` adds `tryTerminate` to the `Terminate` fixture, which catches the revert and returns; the call succeeds with the contract and balances untouched. With the dry run stubbed out, all 8 encumbered and catch cases fail and the 2 controls pass. 1a602ef9d89

File: substrate/frame/revive/uapi/sol/ISystem.sol (hunk 7 above)
Comment (dimartiro) [reply substrate/frame/revive/uapi/sol/ISystem.sol]: Used your wording: "the contract's balance has a lock, freeze or hold that terminate cannot release". 1a602ef9d89

File: prdoc/pr_13085.prdoc (hunk 8 above)
Comment (dimartiro) [reply prdoc/pr_13085.prdoc]: Rewrote it: it says when `terminate` now reverts and that contracts which used to get a success with a live contract now see a catchable revert, plus the `Error` variant, the extra weight and the fixture change for runtime devs. No em dashes. `pallet-revive-fixtures` is listed as `major` since `tryTerminate` adds a variant to the generated `TerminateCalls` enum. 1a602ef9d89

**Review by athei — CHANGES_REQUESTED** [review]
Sorry for the back and forth. The dry run does what I asked for last round, but that turns out to be the wrong ask: it makes a third party's lock revert `System.terminate`. Anyone can put a lock on a contract with a vested transfer of `MinVestedTransfer` (100 CENTS on Asset Hub Westend, also possible from a contract through the vesting precompile) and `per_block = 1`, so it lasts as long as they like. With this PR every `System.terminate` of that contract then reverts, and a contract that calls it without catching the revert can no longer pay out through it. On master the same lock only keeps the contract alive. A terminate that any account can block for 1 WND is not useful, so `terminate` itself has to succeed despite encumbrances it did not create.

What I'd like instead. At call time in `terminate_caller`, after the existing checks:

1. Refund the storage deposit to the termination origin first, as far as the freeze allows, and leave any remainder on hold. The order matters: `evm_balance` is `free - max(frozen - reserved, ED)`, so the hold counts toward the freeze, and after the `Preserve` payout the polite hold release has nothing left to release. The part that can stay held is at most locked funds the contract already paid out, so only the one who placed the lock loses anything.
2. Then do the immediate `Preserve` payout as today. It now leaves exactly the frozen part, or the ED, on the account.
3. Carry the refunded amount in `TerminateArgs` so `do_terminate` still passes it to `transaction_meter.terminate`.

At the end of the call stack `do_terminate` keeps its current order (refund, consumer, ED, sweep, code refcount, deletion) but no step may abort the teardown any more:

4. Refund, best effort, whatever deposit is still held: nothing on the precompile path unless the freeze kept a remainder back, and for `SELFDESTRUCT` the refund it does today. Pass both amounts to `transaction_meter.terminate`.
5. Drop the consumer added at instantiation and burn the ED if the account can be reaped. Otherwise keep the ED and leave the account in place as a plain account. A contract deployed later at the same address takes it over, since instantiation skips `init_contract` for an existing account.
6. Send whatever arrived after the call to the beneficiary, keeping the ED on the account if step 5 kept it. If the transfer fails (the #13039 case), leave the funds on the account.
7. Always delete the contract: `decrement_refcount`, `queue_for_deletion`, `AccountInfoOf` and `ImmutableDataOf`. This must not be skipped once the deposit was refunded at call time, otherwise a live contract keeps storage with no deposit behind it.

The dry run, `TerminateDryRun` and `TerminateBalanceLocked` go away, and the balance line in `ISystem.sol` should be removed rather than reworded. `seal_terminate_logic` has to measure the call-time refund plus the deferred teardown.

`SELFDESTRUCT` (and the PVM `terminate` syscall) has to keep EIP-6780 semantics, including when things happen, so `terminate_if_same_tx` stays exactly as it is: the balance moves when the opcode executes, there is no deposit refund at that point, and the contract is destroyed only if it was created in the same transaction. Only steps 1 to 3 move to call time, and they are for the precompile only. A pre-existing contract keeps its deposits and its code, as on Ethereum, and a lock can never make the opcode fail. It shares only the deferred steps 4 to 7 through `do_terminate`, which still run at the end of the call stack in the order they run today. For a same-transaction contract without an encumbrance they do exactly what they do today. With a lock placed later in that transaction, today the destruction is skipped and the contract survives, which no EVM contract can observe; after the change the contract is deleted and only the locked remainder stays behind, which is as close to EVM behavior as a lock allows.

Changing what `System.terminate` does is fine for a different reason. It is not an EVM opcode but our own precompile, the only way to remove a pre-existing contract, so there are no EVM semantics to keep, only its documented behavior and what deployed callers can observe. When the balance is not encumbered nothing changes for them, apart from the deposit refund arriving at call time instead of at the end of the stack. When it is encumbered, master returns success and leaves the contract alive (#13017), which contradicts both the success and the `ISystem.sol` promise to revert. No deployed contract can depend on that revert, because master never raised it. After the change a success means the contract is gone, which is what callers already assume. The revert this PR adds is the one change deployed callers could actually break on, which is the other reason not to ship it.

Tests: keep the matrix, but the encumbered cases should assert that the call succeeds, the contract is gone, the beneficiary received the transferable part, the locked amount stays on the account and the origin got the deposit back. Please add a case where the contract pays out `address(this).balance` before calling `terminate` under a lock (partial refund), one for a same-transaction `SELFDESTRUCT` under a lock, one showing that `SELFDESTRUCT` of a pre-existing contract under a lock still only moves the balance and keeps the contract and its deposits, and a redeploy at the same address onto a leftover account. The prdoc and the description need rewriting for the new behavior, and the bumps need a fresh look once the error variant is gone.

**Comment by dimartiro** [comment]
Thanks, that makes sense. I reworked the PR along your steps, one commit per part:

- The dry run, `TerminateDryRun` and `TerminateBalanceLocked` are gone, and the balance line in `ISystem.sol` is removed.
- `terminate_caller` refunds the deposit before the `Preserve` payout, as far as the freeze allows. For that, the native `refund_all` now releases `min(hold, reducible_total_balance_on_hold)` instead of failing.
- `do_terminate` keeps its order, but each balance step runs in its own transaction and is skipped on failure. The deletion always runs. `terminate_if_same_tx` is untouched.
- `transaction_meter.terminate` gets the sum of the call-time and the deferred refund. The meter only adds it to the total, so I kept its signature.
- The tests cover your list, plus a lock of 500, which is the case where the refund order matters: with the call-time refund disabled, it is the only one that fails.

Three things you may want to look at:

1. With the refund carried in `TerminateArgs`, terminating twice in one transaction would overwrite the first refund with 0 when frames are merged. The merge now adds up `refunded`, and `precompile_terminate_twice_reports_refund_once` covers it.
2. The #13039 test in `terminate_storage.rs` asserted that the inner contract survives when the sweep fails. It now asserts that the contract is deleted and the funds stay on its account, as in step 6.
3. While writing the same-transaction test I found that a `SELFDESTRUCT` in the constructor of a top-level instantiation does not delete the contract, also on master: the first frame is never added to `contracts_created`. I left it out of this PR and used a nested `new` for the test. Happy to open an issue for it.

`seal_terminate_logic` now measures the call-time refund plus the teardown, so the weights need regenerating. Could you run `/cmd bench` for it?

The prdoc and the description are rewritten. The bumps are now `patch` for `pallet-revive`, since the error variant is gone and nothing public changes, `patch` for `pallet-revive-uapi`, and `major` for `pallet-revive-fixtures` because of the new fixture functions.
